File: Delphi/Conversation/Markdown.swift
import Foundation

/// Text shaping for the stage. Mirrors the web app's markdown util: it strips
/// the model's reasoning/`<situation n>` tags, splits text into paragraphs, and
/// (for assistant replies) renders inline bold/italic.
enum Stage {
    /// Removes reasoning blocks (`<think>…`, possibly unclosed while streaming)
    /// and stray inline tags like `<situation 1>`, then tidies whitespace.
    static func stripTags(_ text: String) -> String {
        var s = text
        // Reasoning blocks — matched to the closing tag or end-of-string.
        s = replacing(
            s,
            pattern: "<(think|thinking|reasoning|reflection|analysis|scratchpad)\\b[^>]*>[\\s\\S]*?(?:</\\1>|$)",
            with: "",
            options: [.caseInsensitive])
        s = replacing(s, pattern: "<[^>]*>", with: "")   // standalone tags
        s = replacing(s, pattern: "<[^>]*$", with: "")    // partial tag still arriving
        s = replacing(s, pattern: "\\n{3,}", with: "\n\n")
        return s.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    /// Blank-line-separated blocks, each trimmed; empty blocks dropped.
    static func paragraphs(_ text: String) -> [String] {
        guard let re = try? NSRegularExpression(pattern: "\\n{2,}") else { return [text] }
        let ns = text as NSString
        var result: [String] = []
        var cursor = 0
        for match in re.matches(in: text, range: NSRange(location: 0, length: ns.length)) {
            result.append(ns.substring(with: NSRange(location: cursor, length: match.range.location - cursor)))
            cursor = match.range.location + match.range.length
        }
        result.append(ns.substring(from: cursor))
        return result
            .map { $0.trimmingCharacters(in: .newlines) }
            .filter { !$0.isEmpty }
    }

    /// Inline markdown (bold/italic) for one paragraph, preserving its single
    /// line breaks. Falls back to plain text if parsing fails.
    static func inline(_ paragraph: String, markdown: Bool) -> AttributedString {
        guard markdown else { return AttributedString(paragraph) }
        let options = AttributedString.MarkdownParsingOptions(
            interpretedSyntax: .inlineOnlyPreservingWhitespace)
        return (try? AttributedString(markdown: paragraph, options: options))
            ?? AttributedString(paragraph)
    }

    // MARK: -

    private static func replacing(
        _ s: String, pattern: String, with template: String,
        options: NSRegularExpression.Options = []
    ) -> String {
        guard let re = try? NSRegularExpression(pattern: pattern, options: options) else { return s }
        let range = NSRange(s.startIndex..., in: s)
        return re.stringByReplacingMatches(in: s, range: range, withTemplate: template)
    }
}

File: web/src/utils/markdown.ts
// Strips reasoning/thinking blocks and inline tags (e.g. <situation N>) from a
// streaming response. Handles unclosed blocks (still streaming) by matching to end-of-string.
const THINK_BLOCK = /<(think|thinking|reasoning|reflection|analysis|scratchpad)\b[^>]*>[\s\S]*?(?:<\/\1>|$)/gi

export const stripAngleTags = (text: string): string =>
  text
    .replace(THINK_BLOCK, '')
    .replace(/<[^>]*>/g, '')   // standalone tags like <situation 1>
    .replace(/<[^>]*$/, '')    // partial tag still arriving
    .replace(/\n{3,}/g, '\n\n')
    .trim()

// Escape HTML, then expand a tiny subset of markdown (bold, italic). XSS-safe because
// we escape before introducing any tags.
export const renderBasicMarkdown = (text: string): string => {
  let html = text
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')

  html = html.replace(/\*\*((?:(?!\n\n)[\s\S])+?)\*\*/g, '<strong>$1</strong>')
  html = html.replace(/\*((?:(?!\n\n)[\s\S])+?)\*/g, '<em>$1</em>')
  html = html.replace(/(?<=^|\W)_((?:(?!\n\n)[\s\S])+?)_(?=$|\W)/g, '<em>$1</em>')

  // Unclosed tags at end of stream
  html = html.replace(/\*\*([^<]*)$/, '<strong>$1</strong>')
  html = html.replace(/\*([^<]*)$/, '<em>$1</em>')
  html = html.replace(/(?<=^|\W)_([^<]*)$/, '<em>$1</em>')

  return html
}
